### PR TITLE
SAAS-3804: Added Access Analyzer Active Findings Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: collectors/aws/accessanalyzer/listFindings.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            if (data.nextToken && data.nextToken.length) {
+                paginating = true;
+                return execute(data.nextToken);
+            }
+
+            cb();
+        };
+
+        function execute(nextToken) { // eslint-disable-line no-inner-declarations
+            var localParams = JSON.parse(JSON.stringify(params || {}));
```
Comment (m-nouman-suleman) [inline collectors/aws/accessanalyzer/listFindings.js:36]: why do we have do this? what is. the issue with params object
```suggestion
            var localParams = params;
```
```suggestion
            var localParams = JSON.parse(JSON.stringify(params || {}));
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                "updatedAt": "2022-01-12T13:48:20+00:00"
+            }
+        ]
+}
+        
+];
+
+
+const createCache = (analyzer, listFindings, analyzerErr, listFindingsErr) => {
+
```
Comment (m-nouman-suleman) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:143]:
```suggestion
```suggestion
```
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                        'Amazon IAM access analyzer have no active findings.',
+                        region, resource);
+                    continue;         
+                } else {
+                    helpers.addResult(results, 2,
+                        'Amazon IAM access analyzer has active findings.',
+                        region, resource);
+                    continue;
+                }
+
```
Comment (m-nouman-suleman) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:65]:
```suggestion
```

File: exports.js [hunk 4]
```diff
@@ -2,491 +2,492 @@
...
+        // 'cloudtrailEncryption'          : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailEncryption.js'),
+        // 'cloudtrailFileValidation'      : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailFileValidation.js'),
+        // 'cloudtrailObjectLock'          : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailObjectLock.js'),
+        // 'cloudtrailToCloudwatch'        : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailToCloudwatch.js'),
+        // 'cloudtrailBucketPrivate'       : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailBucketPrivate.js'),
+        // 'cloudtrailDeliveryFailing'     : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailDeliveryFailing.js'),
+        // 'cloudtrailManagementEvents'    : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailManagementEvents.js'),
+        // 'cloudtrailS3Bucket'            : require(__dirname + '/plugins/aws/cloudtrail/cloudtrailS3Bucket.js'),
+        // 'globalLoggingDuplicated'       : require(__dirname + '/plugins/aws/cloudtrail/globalLoggingDuplicated.js'),
+
```
Comment (m-nouman-suleman) [inline exports.js:68]: Is this intentional?

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                    helpers.addResult(results, 3,
+                        `Unable to get findings description: ${helpers.addError(listFindings)}`,
+                        region, resource);
+                    continue;
+                } 
+                
+                let filtered = listFindings.data.findings.filter(finding => finding.status === 'ACTIVE');
+                if (!filtered.length) {
+                    helpers.addResult(results, 0,
+                        'Amazon IAM access analyzer have no active findings.',
```
Comment (sadeed12345) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:56]:
```suggestion
                        'Amazon IAM access analyzer have no active findings',
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                        region, resource);
+                    continue;         
+                } else {
+                    helpers.addResult(results, 2,
+                        'Amazon IAM access analyzer has active findings.',
```
Comment (sadeed12345) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:61]:
```suggestion
                        'Amazon IAM access analyzer has active findings',
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Access Analyzer Active Findings',
+    category: 'IAM',
+    domain: 'Management and Governance',
+    description: 'Ensure that IAM Access analyzer findings are reviewed for resolving security issues by taking all necessary actions.',
+    more_info: 'Access Analyzer review all of the findings in your account to determine whether the sharing is expected and approved. If the sharing identified in the finding is expected, you can archive the finding.' +
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:9]:
```suggestion
    more_info: 'Access Analyzer review all of the findings in your account to determine whether the sharing is expected and approved. If the sharing identified in the finding is expected, you can archive the finding. ' +
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            }
+
+            for (let analyzer of listAnalyzers.data) {
+                if (!analyzer.arn) continue;
+
+                let resource = analyzer.arn;
+
+                var listFindings = helpers.addSource(cache, source,
+                    ['accessanalyzer', 'listFindings', region, analyzer.arn]);
+
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:44]:
```suggestion
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+                if (!listFindings || listFindings.err || !listFindings.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to get findings description: ${helpers.addError(listFindings)}`,
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:48]:
```suggestion
                        `Unable to query list findings: ${helpers.addError(listFindings)}`,
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+                if (!filtered.length) {
+                    helpers.addResult(results, 0,
+                        'Amazon IAM access analyzer have no active findings',
+                        region, resource);
+                    continue;         
+                } else {
+                    helpers.addResult(results, 2,
+                        'Amazon IAM access analyzer has active findings',
+                        region, resource);
+                    continue;
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:63]:
```suggestion   
                } else {
                    helpers.addResult(results, 2,
                        'Amazon IAM access analyzer has active findings',
                        region, resource);
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+describe('activeFindingsEnabled', function () {
+    describe('run', function () {
+        it('should FAIL if Amazon IAM access analyzer has active findings.', function (done) {
+            const cache = createCache(listAnalyzers, listFindings[0]);
+            activeFindingsEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Amazon IAM access analyzer has active findings.');
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:172]:
```suggestion
                expect(results[0].message).to.include('Amazon IAM access analyzer has active findings');
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+            });
+        });
+
+        it('should PASS if Amazon IAM access analyzer have no active findings.', function (done) {
+            const cache = createCache(listAnalyzers, listFindings[1]);
+            activeFindingsEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Amazon IAM access analyzer have no active findings.');
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:183]:
```suggestion
                expect(results[0].message).to.include('Amazon IAM access analyzer have no active findings');
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('No IAM access analyzers found');
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to list analyzers', function (done) {
+            const cache = createCache(null, null, null, { message: "Unable to list analyzers" });
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:201]:
```suggestion
            const cache = createCache(null, null, { message: "Unable to list analyzers" });
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+module.exports = {
+    title: 'Access Analyzer Active Findings',
+    category: 'IAM',
+    domain: 'Management and Governance',
+    description: 'Ensure that IAM Access analyzer findings are reviewed for resolving security issues by taking all necessary actions.',
+    more_info: 'Access Analyzer review all of the findings in your account to determine whether the sharing is expected and approved. If the sharing identified in the finding is expected, you can archive the finding. ' +
+        'When you archive a finding, the status is changed to Archived, and the finding is removed from the Active findings list. The finding is not deleted. You can view your archived findings at any time. Work through all of the findings in your account until you have zero active findings.',
+    link: 'https://docs.aws.amazon.com/IAM/latest/UserGuide/access-analyzer-work-with-findings.html',
+    recommended_action: 'Enable Access Analyzer to work through all of the findings in your account until you have zero active findings',
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:12]:
```suggestion
    recommended_action: 'Investigate into active findings in your account and do the needful until you have zero active findings.',
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+                if (!analyzer.arn) continue;
+
+                let resource = analyzer.arn;
+
+                var listFindings = helpers.addSource(cache, source,
+                    ['accessanalyzer', 'listFindings', region, analyzer.arn]);
+
+                if (!listFindings || listFindings.err || !listFindings.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query list findings: ${helpers.addError(listFindings)}`,
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:47]:
```suggestion
                        `Unable to IAM Access Analyzer findings: ${helpers.addError(listFindings)}`,
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+            if (!listAnalyzers) return rcb();
+
+            if (listAnalyzers.err || !listAnalyzers.data) {
+                helpers.addResult(results, 3,
+                    `Unable to query for IAM access analyzers: ${helpers.addError(listAnalyzers)}`,region);
+                return rcb();
+            }
+
+            if (!listAnalyzers.data.length) {
+                helpers.addResult(results, 0, 'No IAM access analyzers found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:33]:
```suggestion
                helpers.addResult(results, 0, 'No IAM Access Analyzer analyzers found', region);
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+                    helpers.addResult(results, 3,
+                        `Unable to IAM Access Analyzer findings: ${helpers.addError(listFindings)}`,
+                        region, resource);
+                    continue;
+                } 
+                
+                let filtered = listFindings.data.findings.filter(finding => finding.status === 'ACTIVE');
+                if (!filtered.length) {
+                    helpers.addResult(results, 0,
+                        'Amazon IAM access analyzer have no active findings',
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:55]:
```suggestion
                        'Amazon IAM Access Analyzer have no active findings',
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+                        region, resource);
+                } else {
+                    helpers.addResult(results, 2,
+                        'Amazon IAM access analyzer has active findings',
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:59]:
```suggestion
                        'Amazon IAM Access Analyzer has active findings',
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+
+describe('activeFindingsEnabled', function () {
+    describe('run', function () {
+        it('should FAIL if Amazon IAM access analyzer has active findings.', function (done) {
+            const cache = createCache(listAnalyzers, listFindings[0]);
+            activeFindingsEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Amazon IAM access analyzer has active findings');
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:172]:
```suggestion
                expect(results[0].message).to.include('Amazon IAM Access Analyzer has active findings');
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+            });
+        });
+
+        it('should PASS if Amazon IAM access analyzer have no active findings.', function (done) {
+            const cache = createCache(listAnalyzers, listFindings[1]);
+            activeFindingsEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Amazon IAM access analyzer have no active findings');
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:183]:
```suggestion
                expect(results[0].message).to.include('Amazon IAM Access Analyzer has no active findings');
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.js [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+                    helpers.addResult(results, 3,
+                        `Unable to IAM Access Analyzer findings: ${helpers.addError(listFindings)}`,
+                        region, resource);
+                    continue;
+                } 
+                
+                let filtered = listFindings.data.findings.filter(finding => finding.status === 'ACTIVE');
+                if (!filtered.length) {
+                    helpers.addResult(results, 0,
+                        'Amazon IAM Access Analyzer have no active findings',
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.js:55]:
```suggestion
                        'Amazon IAM Access Analyzer has no active findings',
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+            });
+        });
+
+        it('should PASS if no analyzers found', function (done) {
+            const cache = createCache([]);
+            activeFindingsEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('No IAM access analyzers found');
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:195]:
```suggestion
                expect(results[0].message).to.include('No IAM Access Analyzer analyzers found');
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+                done();
+            });
+        });
+
+        it('should UNKNOWN if Unable to query for IAM access analyzers', function (done) {
+            const cache = createCache(null, null, { message: "Unable to query for IAM access analyzers" });
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:201]:
```suggestion
            const cache = createCache(null, null, { message: "Unable to query for IAM Access Analyzer analyzers" });
```

File: plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+            activeFindingsEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Unable to query for IAM access analyzers');
```
Comment (AkhtarAmir) [inline plugins/aws/accessanalyzer/activeFindingsEnabled.spec.js:206]:
```suggestion
                expect(results[0].message).to.include('Unable to query for IAM Access Analyzer analyzers');
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**